app/models/minion.rb

```diff
@@ -9,37 +9,69 @@ class Minion < ApplicationRecord
   class NotEnoughMinions < StandardError; end
   # Raised when we fail to assign a role on a minion
   class CouldNotAssignRole < StandardError; end
+  # Raised when Minion doesn't exist
+  class NonExistingMinion < StandardError; end
 
   enum highstate: [:not_applied, :pending, :failed, :applied]
   enum role: [:master, :minion]
 
-  validates :hostname, uniqueness: true
+  validates :hostname, presence: true, uniqueness: true
 
-  # This method is used to assign the specified roles to Minions which do not
-  # have a roles already assigned (available minions).
-  # roles param are the roles we absolutely have to assign. If we can't assign
-  # one of those, the method will raise.
-  # default_role param can be set if we want all the rest of the available
-  # minions to get a default role.
-  # Returns the ids of the minions on which a roles has been assigned.
-  def self.assign_roles(roles: [], default_role: nil)
-    minions = Minion.where(role: nil)
-
-    # only the needed number of minions or all if we have a default role
-    minions = minions.limit(roles.size) unless default_role
-
-    raise NotEnoughMinions if minions.count < roles.size
+  # Example:
+  #   Minion.assign_roles(
+  #     roles: {
+  #       master: ["master.example.com"],
+  #       minion: ["minion1.example.com"]
+  #     },
+  #     default_role: :dns
+  #   )
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  def self.assign_roles(roles: {}, default_role: :minion)
+    if roles[:master].any? && Minion.find_by(hostname: roles[:master].first).nil?
+      raise NonExistingMinion, "Failed to process non existing minion: #{roles[:master].first}"
+    end
+    master = Minion.find_by(hostname: roles[:master].first)
+    # choose requested minions or all other than master
+    minions = if roles[:minion]
+      Minion.where(hostname: roles[:minion])
+    else
+      Minion.where(role: nil).where.not(hostname: roles[:master].first)
+    end
```

Comment:
Can we take a more generic approach? As in, not check each roles separately but iterate over the ```roles```. E.g.

```
assigned_ids = []
roles.each do |role, minion_ids|
  minions = Minion.where(hostname: minion_ids).each do |minion|
    if minion.assign_role(role)
      assigned_ids << minion.id
    else
      raise CouldNotAssignRole, "Failed to assign #{role} role to #{minion.hostname}"
    end
  end
end
```
(Warning: not tested code)